src/Symfony/Component/DependencyInjection/ContainerBuilder.php

```diff
@@ -1050,13 +1051,48 @@ private function createService(Definition $definition, array &$inlineServices, b
         }
 
         $parameterBag = $this->getParameterBag();
+        $class = ($parameterBag->resolveValue($definition->getClass()) ?: (['Closure', 'fromCallable'] === $definition->getFactory() ? 'Closure' : null));
 
-        if (true === $tryProxy && $definition->isLazy() && !$tryProxy = !($proxy = $this->proxyInstantiator ??= new LazyServiceInstantiator()) || $proxy instanceof RealServiceInstantiator) {
+        if ('Closure' === $class && $definition->isLazy() && ['Closure', 'fromCallable'] === $definition->getFactory()) {
+            $callable = $parameterBag->unescapeValue($parameterBag->resolveValue($definition->getArgument(0)));
+
+            if ($callable instanceof Reference || $callable instanceof Definition) {
+                $callable = [$callable, '__invoke'];
+            }
+
+            if (\is_array($callable) && (
+                $callable[0] instanceof Reference
+                || $callable[0] instanceof Definition && !isset($inlineServices[spl_object_hash($callable[0])])
+            )) {
+                $containerRef = $this->containerRef ??= \WeakReference::create($this);
+                $class = ($callable[0] instanceof Reference ? $this->findDefinition($callable[0]) : $callable[0])->getClass();
+                $r = $this->getReflectionClass($class);
+
+                if ($r && $r->hasMethod($callable[1])) {
+                    $signature = ProxyHelper::exportSignature($r->getMethod($callable[1]));
+                    $signature = preg_replace('/[^(]++(\(.*\))(: .*?)?$/', '\1 use ($containerRef, $callable, &$inlineServices)', $signature);
```

Comment:
What happens if the method being proxied has a parameter named `$callable` or `$containerRef` ?

Reply:
good catch, fixed